File: app/routers/employees.py
from datetime import date, datetime
from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import or_
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app import models, schemas
from app.database import get_db
from app.dependencies import confirm_critical_action, require_permission
from app.services import audit


router = APIRouter(prefix="/funcionarios", tags=["Funcionários"])
employee_access = require_permission("pode_gerenciar_funcionarios")


def _mask_cpf(value: str) -> str:
    return f"***.***.***-{value[-2:]}"


def _mask_email(value: str | None) -> str | None:
    if not value:
        return None
    local, domain = value.split("@", 1)
    return f"{local[:2]}***@{domain}"


def _mask_phone(value: str | None) -> str | None:
    return f"*******{value[-4:]}" if value else None


def _audit_snapshot(employee: models.Funcionario) -> dict:
    return {
        "matricula": employee.matricula,
        "nome_completo": employee.nome_completo,
        "cpf": _mask_cpf(employee.cpf),
        "rg_final": f"***{employee.rg[-4:]}",
        "email_pessoal": _mask_email(employee.email_pessoal),
        "celular": _mask_phone(employee.celular),
        "departamento": employee.departamento,
        "cargo": employee.cargo,
        "tipo_contrato": employee.tipo_contrato,
        "data_admissao": employee.data_admissao,
        "status": employee.status,
        "data_desligamento": employee.data_desligamento,
        "usuario_id": employee.usuario_id,
    }


def _summary(employee: models.Funcionario) -> dict:
    user = employee.usuario
    return {
        "id": employee.id,
        "matricula": employee.matricula,
        "nome_completo": employee.nome_completo,
        "nome_social": employee.nome_social,
        "cpf_mascarado": _mask_cpf(employee.cpf),
        "celular": employee.celular,
        "email_corporativo": employee.email_corporativo,
        "departamento": employee.departamento,
        "cargo": employee.cargo,
        "tipo_contrato": employee.tipo_contrato,
        "status": employee.status,
        "data_admissao": employee.data_admissao,
        "data_desligamento": employee.data_desligamento,
        "usuario_id": employee.usuario_id,
        "usuario_login": user.usuario_login if user else None,
        "usuario_ativo": user.ativo if user else None,
        "atualizado_em": employee.atualizado_em,
    }


def _detail(employee: models.Funcionario) -> dict:
    data = {
        field: getattr(employee, field)
        for field in schemas.FuncionarioBase.model_fields
    }
    user = employee.usuario
    return {
        **data,
        "id": employee.id,
        "matricula": employee.matricula,
        "usuario_nome": user.nome if user else None,
        "usuario_login": user.usuario_login if user else None,
        "usuario_ativo": user.ativo if user else None,
        "criado_em": employee.criado_em,
        "atualizado_em": employee.atualizado_em,
        "criado_por_nome": employee.criado_por_nome,
        "atualizado_por_nome": employee.atualizado_por_nome,
    }


def _employee_or_404(db: Session, employee_id: int) -> models.Funcionario:
    employee = db.get(models.Funcionario, employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Funcionário não encontrado")
    return employee


def _validate_link(
    db: Session,
    user_id: int | None,
    employee_id: int | None = None,
) -> models.Usuario | None:
    if not user_id:
        return None
    user = db.get(models.Usuario, user_id)
    if not user:
        raise HTTPException(status_code=400, detail="Conta de acesso não encontrada")
    linked = db.query(models.Funcionario).filter(
        models.Funcionario.usuario_id == user_id,
        models.Funcionario.id != employee_id if employee_id else True,
    ).first()
    if linked:
        raise HTTPException(
            status_code=409,
            detail=f"Esta conta já está vinculada ao funcionário {linked.nome_completo}",
        )
    return user


def _validate_uniques(
    db: Session,
    data: dict,
    employee_id: int | None = None,
) -> None:
    checks = (
        ("cpf", data.get("cpf"), "CPF"),
        ("matricula", data.get("matricula"), "matrícula"),
        ("pis_pasep", data.get("pis_pasep"), "PIS/PASEP"),
    )
    for field, value, label in checks:
        if not value:
            continue
        query = db.query(models.Funcionario).filter(getattr(models.Funcionario, field) == value)
        if employee_id:
            query = query.filter(models.Funcionario.id != employee_id)
        if query.first():
            raise HTTPException(status_code=409, detail=f"Já existe um funcionário com este {label}")


def _sync_access(employee: models.Funcionario, user: models.Usuario | None) -> None:
    if not user:
        return
    user.nome = employee.nome_completo
    user.email = employee.email_corporativo or employee.email_pessoal
    user.telefone = employee.celular
    if employee.status == "DESLIGADO":
        user.ativo = False
        user.tentativas_login = 0
        user.bloqueado_ate = None


def _require_permanent_delete(
    user: models.Usuario = Depends(employee_access),
) -> models.Usuario:
    if user.tipo_usuario not in {"DONO", "DESENVOLVEDOR"}:
        raise HTTPException(
            status_code=403,
            detail="Exclusão definitiva permitida somente para Dono ou Desenvolvedor",
        )
    return user


@router.get("/resumo")
def employee_overview(
    db: Session = Depends(get_db),
    _: models.Usuario = Depends(employee_access),
):
    active = db.query(models.Funcionario).filter(models.Funcionario.status == "ATIVO").count()
    terminated = db.query(models.Funcionario).filter(models.Funcionario.status == "DESLIGADO").count()
    departments = [
        item[0]
        for item in db.query(models.Funcionario.departamento)
        .distinct()
        .order_by(models.Funcionario.departamento)
        .all()
    ]
    return {
        "total": active + terminated,
        "ativos": active,
        "desligados": terminated,
        "departamentos": departments,
    }


@router.get("/usuarios-disponiveis", response_model=List[schemas.UsuarioVinculavelResponse])
def linkable_users(
    db: Session = Depends(get_db),
    _: models.Usuario = Depends(employee_access),
):
    linked = {
        item.usuario_id: item.id
        for item in db.query(models.Funcionario).filter(models.Funcionario.usuario_id.is_not(None)).all()
    }
    return [
        {
            "id": user.id,
            "nome": user.nome,
            "usuario_login": user.usuario_login,
            "ativo": user.ativo,
            "funcionario_id": linked.get(user.id),
        }
        for user in db.query(models.Usuario).order_by(models.Usuario.nome).all()
    ]


@router.get("/", response_model=List[schemas.FuncionarioResumo])
def list_employees(
    busca: str | None = Query(default=None, max_length=160),
    status: str | None = Query(default=None, pattern="^(ATIVO|DESLIGADO)$"),
    departamento: str | None = Query(default=None, max_length=120),
    db: Session = Depends(get_db),
    _: models.Usuario = Depends(employee_access),
):
    query = db.query(models.Funcionario)
    if busca and busca.strip():
        term = busca.strip()
        like = f"%{term}%"
        digits = "".join(character for character in term if character.isdigit())
        conditions = [
            models.Funcionario.nome_completo.ilike(like),
            models.Funcionario.nome_social.ilike(like),
            models.Funcionario.matricula.ilike(like),
            models.Funcionario.email_pessoal.ilike(like),
            models.Funcionario.email_corporativo.ilike(like),
            models.Funcionario.cargo.ilike(like),
            models.Funcionario.departamento.ilike(like),
        ]
        if digits:
            conditions.append(models.Funcionario.cpf.ilike(f"%{digits}%"))
        query = query.filter(or_(*conditions))
    if status:
        query = query.filter(models.Funcionario.status == status)
    if departamento:
        query = query.filter(models.Funcionario.departamento == departamento.strip())
    employees = query.order_by(models.Funcionario.nome_completo).limit(1000).all()
    return [_summary(employee) for employee in employees]


@router.get("/{employee_id}", response_model=schemas.FuncionarioResponse)
def get_employee(
    employee_id: int,
    db: Session = Depends(get_db),
    _: models.Usuario = Depends(employee_access),
):
    return _detail(_employee_or_404(db, employee_id))


@router.post("/", response_model=schemas.FuncionarioResponse, status_code=201)
def create_employee(
    data: schemas.FuncionarioCreate,
    db: Session = Depends(get_db),
    actor: models.Usuario = Depends(employee_access),
):
    values = data.model_dump()
    _validate_uniques(db, values)
    linked_user = _validate_link(db, values.get("usuario_id"))
    if data.status == "DESLIGADO" and data.usuario_id == actor.id:
        raise HTTPException(status_code=400, detail="Você não pode desligar o próprio acesso")
    employee = models.Funcionario(
        **values,
        criado_por_id=actor.id,
        criado_por_nome=actor.nome,
        atualizado_por_id=actor.id,
        atualizado_por_nome=actor.nome,
    )
    db.add(employee)
    try:
        db.flush()
        if not employee.matricula:
            employee.matricula = f"FUNC-{employee.id:05d}"
        _sync_access(employee, linked_user)
        audit(
            db, actor, "FUNCIONARIOS", "CRIADO", "funcionarios", employee.id,
            after=_audit_snapshot(employee),
        )
        db.commit()
    except IntegrityError as error:
        db.rollback()
        raise HTTPException(status_code=409, detail="CPF, matrícula, PIS/PASEP ou conta já cadastrados") from error
    db.refresh(employee)
    return _detail(employee)


@router.put("/{employee_id}", response_model=schemas.FuncionarioResponse)
def update_employee(
    employee_id: int,
    data: schemas.FuncionarioUpdate,
    db: Session = Depends(get_db),
    actor: models.Usuario = Depends(employee_access),
):
    employee = _employee_or_404(db, employee_id)
    values = data.model_dump()
    _validate_uniques(db, values, employee_id)
    linked_user = _validate_link(db, values.get("usuario_id"), employee_id)
    if data.status == "DESLIGADO" and data.usuario_id == actor.id:
        raise HTTPException(status_code=400, detail="Você não pode desligar o próprio acesso")
    before = _audit_snapshot(employee)
    previous_status = employee.status
    for field, value in values.items():
        setattr(employee, field, value)
    employee.atualizado_em = datetime.now()
    employee.atualizado_por_id = actor.id
    employee.atualizado_por_nome = actor.nome
    _sync_access(employee, linked_user)
    action = "ATUALIZADO"
    if previous_status != employee.status:
        action = "DESLIGADO" if employee.status == "DESLIGADO" else "REATIVADO"
    audit(
        db, actor, "FUNCIONARIOS", action, "funcionarios", employee.id,
        before=before, after=_audit_snapshot(employee),
    )
    try:
        db.commit()
    except IntegrityError as error:
        db.rollback()
        raise HTTPException(status_code=409, detail="CPF, matrícula, PIS/PASEP ou conta já cadastrados") from error
    db.refresh(employee)
    return _detail(employee)


@router.post("/{employee_id}/status", response_model=schemas.FuncionarioResponse)
def change_employee_status(
    employee_id: int,
    data: schemas.FuncionarioStatusUpdate,
    db: Session = Depends(get_db),
    actor: models.Usuario = Depends(employee_access),
):
    employee = _employee_or_404(db, employee_id)
    if data.status == "DESLIGADO" and employee.usuario_id == actor.id:
        raise HTTPException(status_code=400, detail="Você não pode desligar o próprio acesso")
    before = _audit_snapshot(employee)
    if data.status == "DESLIGADO":
        termination_date = data.data_desligamento or date.today()
        if termination_date < employee.data_admissao:
            raise HTTPException(status_code=400, detail="Desligamento não pode ser anterior à admissão")
        if termination_date > date.today():
            raise HTTPException(status_code=400, detail="Desligamento não pode ter data futura")
        employee.status = "DESLIGADO"
        employee.data_desligamento = termination_date
        employee.motivo_desligamento = data.motivo.strip()
        action = "DESLIGADO"
    else:
        employee.status = "ATIVO"
        employee.data_desligamento = None
        employee.motivo_desligamento = None
        action = "REATIVADO"
    employee.atualizado_em = datetime.now()
    employee.atualizado_por_id = actor.id
    employee.atualizado_por_nome = actor.nome
    _sync_access(employee, employee.usuario)
    audit(
        db, actor, "FUNCIONARIOS", action, "funcionarios", employee.id,
        before=before, after=_audit_snapshot(employee),
    )
    db.commit()
    db.refresh(employee)
    return _detail(employee)


@router.post("/{employee_id}/excluir-definitivamente")
def permanently_delete_employee(
    employee_id: int,
    data: schemas.ConfirmacaoCritica,
    db: Session = Depends(get_db),
    actor: models.Usuario = Depends(_require_permanent_delete),
):
    confirm_critical_action(actor, data.senha)
    employee = _employee_or_404(db, employee_id)
    if employee.status != "DESLIGADO":
        raise HTTPException(
            status_code=409,
            detail="Desligue o funcionário antes da exclusão definitiva",
        )
    before = _audit_snapshot(employee)
    audit(
        db, actor, "FUNCIONARIOS", "EXCLUIDO_DEFINITIVAMENTE", "funcionarios",
        employee.id, before=before, after={"motivo": data.motivo.strip()},
    )
    db.delete(employee)
    db.commit()
    return {"status": "success"}

File: static/js/employees.js
import {$,api,state,esc,money,date,toast,showError,criticalConfirmation,payload} from './api.js?v=5.4.0';

const fields = {
  matricula:'employee_matricula',nome_completo:'employee_nome',nome_social:'employee_nome_social',
  cpf:'employee_cpf',rg:'employee_rg',orgao_emissor_rg:'employee_rg_orgao',uf_rg:'employee_rg_uf',
  data_nascimento:'employee_birth',email_pessoal:'employee_email',email_corporativo:'employee_corporate_email',
  celular:'employee_cell',telefone:'employee_phone',cep:'employee_cep',logradouro:'employee_street',
  numero:'employee_number',complemento:'employee_complement',bairro:'employee_neighborhood',
  cidade:'employee_city',uf:'employee_state',pis_pasep:'employee_pis',ctps_numero:'employee_ctps',
  ctps_serie:'employee_ctps_series',ctps_uf:'employee_ctps_state',departamento:'employee_department',
  cargo:'employee_job',tipo_contrato:'employee_contract',data_admissao:'employee_admission',
  salario_base:'employee_salary',jornada_semanal:'employee_hours',gestor:'employee_manager',
  contato_emergencia_nome:'employee_emergency_name',contato_emergencia_parentesco:'employee_emergency_relation',
  contato_emergencia_telefone:'employee_emergency_phone',status:'employee_status',
  data_desligamento:'employee_termination_date',motivo_desligamento:'employee_termination_reason',
  observacoes:'employee_notes',
};

const optionalFields = new Set([
  'matricula','nome_social','orgao_emissor_rg','uf_rg','email_corporativo','telefone','complemento',
  'pis_pasep','ctps_numero','ctps_serie','ctps_uf','salario_base','jornada_semanal','gestor',
  'data_desligamento','motivo_desligamento','observacoes',
]);

function localToday(){const now=new Date();return `${now.getFullYear()}-${String(now.getMonth()+1).padStart(2,'0')}-${String(now.getDate()).padStart(2,'0')}`}
function phone(value){const digits=String(value||'').replace(/\D/g,'');if(digits.length===11)return `(${digits.slice(0,2)}) ${digits.slice(2,7)}-${digits.slice(7)}`;if(digits.length===10)return `(${digits.slice(0,2)}) ${digits.slice(2,6)}-${digits.slice(6)}`;return value||'—'}
function cpf(value){const digits=String(value||'').replace(/\D/g,'');return digits.length===11?`${digits.slice(0,3)}.${digits.slice(3,6)}.${digits.slice(6,9)}-${digits.slice(9)}`:value||'—'}
function cep(value){const digits=String(value||'').replace(/\D/g,'');return digits.length===8?`${digits.slice(0,5)}-${digits.slice(5)}`:value||'—'}
function textOrNull(id){const value=$(id).value.trim();return value||null}
function numberOrNull(id){return $(id).value===''?null:Number($(id).value)}
function contractName(value){return {CLT:'CLT',PJ:'PJ',ESTAGIO:'Estágio',TEMPORARIO:'Temporário',APRENDIZ:'Aprendiz',OUTRO:'Outro'}[value]||value}

function syncTerminationFields(){
  const terminated=$('employee_status').value==='DESLIGADO';
  $('employeeTerminationFields').classList.toggle('hidden',!terminated);
  $('employee_termination_date').required=terminated;
  $('employee_termination_reason').required=terminated;
  if(terminated&&!$('employee_termination_date').value)$('employee_termination_date').value=localToday();
  if(!terminated){$('employee_termination_date').value='';$('employee_termination_reason').value=''}
}

function employeePayload(){
  const data={};
  Object.entries(fields).forEach(([key,id])=>{data[key]=optionalFields.has(key)?textOrNull(id):$(id).value.trim()});
  data.usuario_id=$('employee_user').value?Number($('employee_user').value):null;
  data.salario_base=numberOrNull('employee_salary');
  data.jornada_semanal=numberOrNull('employee_hours');
  if(data.status!=='DESLIGADO'){data.data_desligamento=null;data.motivo_desligamento=null}
  return data;
}

function renderUserOptions(employeeId=null,selectedUserId=null){
  $('employee_user').innerHTML='<option value="">Sem acesso ao ERP</option>'+state.employeeUsers.map(user=>{
    const used=user.funcionario_id&&user.funcionario_id!==employeeId;
    const selected=user.id===Number(selectedUserId);
    return `<option value="${user.id}" ${used?'disabled':''} ${selected?'selected':''}>${esc(user.nome)} · ${esc(user.usuario_login)}${user.ativo?'':' · inativo'}${used?' · já vinculado':''}</option>`;
  }).join('');
}

function resetEmployeeForm(){
  $('employeeForm').reset();$('employee_id').value='';$('employeeFormTitle').textContent='Novo funcionário';
  $('employee_status').value='ATIVO';renderUserOptions();syncTerminationFields();$('employeePanel').classList.add('hidden');
}

function fillEmployeeForm(employee){
  $('employeeForm').reset();$('employee_id').value=employee.id;$('employeeFormTitle').textContent=`Editar ${employee.nome_completo}`;
  Object.entries(fields).forEach(([key,id])=>{$(id).value=employee[key]??''});
  renderUserOptions(employee.id,employee.usuario_id);syncTerminationFields();$('employeePanel').classList.remove('hidden');
  $('employeePanel').scrollIntoView({behavior:'smooth',block:'start'});
}

async function editEmployee(employeeId){fillEmployeeForm(await api(`/funcionarios/${employeeId}`))}

function renderDepartmentFilter(departments){
  const selected=$('employee_filter_department').value;
  $('employee_filter_department').innerHTML='<option value="">Todos os departamentos</option>'+departments.map(item=>`<option value="${esc(item)}">${esc(item)}</option>`).join('');
  $('employee_filter_department').value=selected;
}

function renderEmployees(list){
  const permanent=['DONO','DESENVOLVEDOR'].includes(state.user.tipo_usuario);
  $('employeeList').innerHTML=list.map(employee=>{
    const active=employee.status==='ATIVO';
    const access=employee.usuario_id?`<span class="badge ${employee.usuario_ativo?'badge-green':'badge-red'}"><i class="fa-solid fa-key"></i> ${employee.usuario_ativo?'Acesso ativo':'Acesso bloqueado'}</span>`:'<span class="badge">Sem acesso ao ERP</span>';
    return `<article class="list-card employee-card ${active?'':'employee-terminated'}"><div class="employee-avatar"><i class="fa-solid ${active?'fa-user-tie':'fa-user-slash'}"></i></div><div class="employee-card-body"><div class="employee-card-title"><div><b>${esc(employee.nome_social||employee.nome_completo)}</b>${employee.nome_social?`<small>${esc(employee.nome_completo)}</small>`:''}</div><span class="badge ${active?'badge-green':'badge-red'}">${active?'Ativo':'Desligado'}</span></div><p>${esc(employee.matricula)} · ${esc(employee.cargo)} · ${esc(employee.departamento)}</p><div class="employee-card-meta"><span><i class="fa-regular fa-id-card"></i> ${esc(employee.cpf_mascarado)}</span><span><i class="fa-solid fa-mobile-screen"></i> ${esc(phone(employee.celular))}</span><span><i class="fa-regular fa-envelope"></i> ${esc(employee.email_corporativo||'Sem e-mail corporativo')}</span><span><i class="fa-regular fa-calendar"></i> Admissão: ${date(employee.data_admissao)}</span></div><div class="employee-access">${access}${employee.usuario_login?`<small>${esc(employee.usuario_login)}</small>`:''}</div></div><div class="actions employee-actions"><button class="btn-secondary" data-view-employee="${employee.id}"><i class="fa-solid fa-eye"></i> Visualizar</button><button class="btn-secondary" data-edit-employee="${employee.id}"><i class="fa-solid fa-pen"></i> Editar</button>${active?`<button class="btn-danger" data-terminate-employee="${employee.id}"><i class="fa-solid fa-user-slash"></i> Desligar</button>`:`<button class="btn-secondary" data-reactivate-employee="${employee.id}"><i class="fa-solid fa-user-check"></i> Reativar</button>${permanent?`<button class="btn-danger" data-delete-employee="${employee.id}"><i class="fa-solid fa-trash"></i> Excluir</button>`:''}`}</div></article>`;
  }).join('')||'<div class="card p-5"><b>Nenhum funcionário encontrado.</b><p class="help-text">Revise os filtros ou adicione o primeiro cadastro.</p></div>';
}

export async function loadEmployees(){
  if(!state.user.pode_gerenciar_funcionarios)return;
  const params=new URLSearchParams();
  if($('employee_search').value)params.set('busca',$('employee_search').value);
  if($('employee_filter_status').value)params.set('status',$('employee_filter_status').value);
  if($('employee_filter_department').value)params.set('departamento',$('employee_filter_department').value);
  const [list,overview,users]=await Promise.all([
    api(`/funcionarios/?${params}`),api('/funcionarios/resumo'),api('/funcionarios/usuarios-disponiveis'),
  ]);
  state.employees=list;state.employeeUsers=users;
  $('employeeTotal').textContent=overview.total;$('employeeActive').textContent=overview.ativos;$('employeeTerminated').textContent=overview.desligados;
  renderDepartmentFilter(overview.departamentos);renderUserOptions(Number($('employee_id').value)||null,$('employee_user').value||null);renderEmployees(list);
}

function detailItem(label,value){return `<div class="employee-detail-item"><span>${esc(label)}</span><b>${esc(value??'—')}</b></div>`}
function detailSection(title,icon,items){return `<section><h3><i class="fa-solid ${icon}"></i>${esc(title)}</h3><div class="employee-detail-grid">${items.join('')}</div></section>`}

async function showEmployee(employeeId){
  const employee=await api(`/funcionarios/${employeeId}`);state.selectedEmployee=employee;
  $('employeeDetailTitle').textContent=employee.nome_social||employee.nome_completo;
  $('employeeDetailSubtitle').textContent=`${employee.matricula} · ${employee.cargo} · ${employee.status==='ATIVO'?'Ativo':'Desligado'}`;
  $('employeeDetailBody').innerHTML=[
    detailSection('Identificação','fa-id-card',[
      detailItem('Nome completo',employee.nome_completo),detailItem('Nome social',employee.nome_social),detailItem('CPF',cpf(employee.cpf)),detailItem('RG',`${employee.rg}${employee.orgao_emissor_rg?` · ${employee.orgao_emissor_rg}`:''}${employee.uf_rg?`/${employee.uf_rg}`:''}`),detailItem('Nascimento',date(employee.data_nascimento)),
    ]),
    detailSection('Contato e endereço','fa-location-dot',[
      detailItem('Celular',phone(employee.celular)),detailItem('Telefone',phone(employee.telefone)),detailItem('E-mail pessoal',employee.email_pessoal),detailItem('E-mail corporativo',employee.email_corporativo),detailItem('Endereço',`${employee.logradouro}, ${employee.numero}${employee.complemento?` · ${employee.complemento}`:''}`),detailItem('Bairro',employee.bairro),detailItem('Cidade/UF',`${employee.cidade}/${employee.uf}`),detailItem('CEP',cep(employee.cep)),
    ]),
    detailSection('Vínculo profissional','fa-briefcase',[
      detailItem('Departamento',employee.departamento),detailItem('Cargo',employee.cargo),detailItem('Contrato',contractName(employee.tipo_contrato)),detailItem('Admissão',date(employee.data_admissao)),detailItem('Salário base',employee.salario_base===null?'Não informado':money(employee.salario_base)),detailItem('Jornada semanal',employee.jornada_semanal===null?'Não informada':`${employee.jornada_semanal} h`),detailItem('Gestor',employee.gestor),detailItem('PIS/PASEP',employee.pis_pasep),detailItem('CTPS',[employee.ctps_numero,employee.ctps_serie,employee.ctps_uf].filter(Boolean).join(' · ')||'—'),detailItem('Conta no ERP',employee.usuario_login?`${employee.usuario_nome} · ${employee.usuario_login} · ${employee.usuario_ativo?'ativa':'bloqueada'}`:'Sem conta vinculada'),
    ]),
    detailSection('Emergência e situação','fa-shield-heart',[
      detailItem('Contato',employee.contato_emergencia_nome),detailItem('Relação',employee.contato_emergencia_parentesco),detailItem('Telefone',phone(employee.contato_emergencia_telefone)),detailItem('Status',employee.status==='ATIVO'?'Ativo':'Desligado'),detailItem('Data do desligamento',date(employee.data_desligamento)),detailItem('Motivo',employee.motivo_desligamento),detailItem('Observações',employee.observacoes),detailItem('Última atualização',`${new Date(employee.atualizado_em).toLocaleString('pt-BR')} · ${employee.atualizado_por_nome}`),
    ]),
  ].join('');
  $('employeeDetailDialog').showModal();
}

export function initEmployees(){
  $('newEmployee').addEventListener('click',()=>{resetEmployeeForm();$('employeePanel').classList.remove('hidden');$('employeePanel').scrollIntoView({behavior:'smooth',block:'start'})});
  $('closeEmployeePanel').addEventListener('click',resetEmployeeForm);$('cancelEmployeeEdit').addEventListener('click',resetEmployeeForm);
  $('employee_status').addEventListener('change',syncTerminationFields);
  $('employeeFilters').addEventListener('submit',event=>{event.preventDefault();loadEmployees().catch(showError)});
  $('clearEmployeeFilters').addEventListener('click',()=>{$('employeeFilters').reset();loadEmployees().catch(showError)});
  $('employeeForm').addEventListener('submit',async event=>{event.preventDefault();try{const id=$('employee_id').value;await api(id?`/funcionarios/${id}`:'/funcionarios/',{method:id?'PUT':'POST',body:JSON.stringify(employeePayload())});toast(id?'Cadastro do funcionário atualizado':'Funcionário cadastrado');resetEmployeeForm();await loadEmployees()}catch(error){showError(error)}});
  $('employeeList').addEventListener('click',async event=>{const view=event.target.closest('[data-view-employee]'),edit=event.target.closest('[data-edit-employee]'),terminate=event.target.closest('[data-terminate-employee]'),reactivate=event.target.closest('[data-reactivate-employee]'),remove=event.target.closest('[data-delete-employee]');try{if(view){await showEmployee(view.dataset.viewEmployee);return}if(edit){await editEmployee(edit.dataset.editEmployee);return}if(terminate){$('employee_status_id').value=terminate.dataset.terminateEmployee;$('employee_status_date').value=localToday();$('employee_status_reason').value='';$('employeeStatusDialog').showModal();return}if(reactivate&&confirm('Reativar este cadastro? A conta de acesso continuará bloqueada até um administrador reativá-la na aba Usuários.')){await api(`/funcionarios/${reactivate.dataset.reactivateEmployee}/status`,payload({status:'ATIVO'}));toast('Cadastro reativado; o acesso ao ERP não foi liberado automaticamente');await loadEmployees();return}if(remove){const confirmation=await criticalConfirmation('Excluir funcionário definitivamente','Use apenas para cadastro duplicado ou criado por engano. O histórico da ação será preservado e a conta de acesso não será excluída.');if(confirmation){await api(`/funcionarios/${remove.dataset.deleteEmployee}/excluir-definitivamente`,payload(confirmation));toast('Cadastro excluído definitivamente');await loadEmployees()}}}catch(error){showError(error)}});
  $('employeeStatusForm').addEventListener('submit',async event=>{event.preventDefault();try{await api(`/funcionarios/${$('employee_status_id').value}/status`,payload({status:'DESLIGADO',data_desligamento:$('employee_status_date').value,motivo:$('employee_status_reason').value}));$('employeeStatusDialog').close();toast('Funcionário desligado; histórico preservado e acesso bloqueado');await loadEmployees()}catch(error){showError(error)}});
  $('cancelEmployeeStatus').addEventListener('click',()=>$('employeeStatusDialog').close());
  $('closeEmployeeDetail').addEventListener('click',()=>$('employeeDetailDialog').close());
  $('editEmployeeFromDetail').addEventListener('click',()=>{const employee=state.selectedEmployee;if(!employee)return;$('employeeDetailDialog').close();fillEmployeeForm(employee)});
  syncTerminationFields();
}
